pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -253,43 +253,59 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
 		return fmt.Errorf("node %s doesn't have providerID", nodeName)
 	}
 
-	cidrs, err := ca.cloud.AliasRangesByProviderID(node.Spec.ProviderID)
+	podCIDRs, err := ca.cloud.AliasRangesByProviderID(node.Spec.ProviderID)
 	if err != nil {
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: %v", err)
 	}
-	if len(cidrs) == 0 {
+	if len(podCIDRs) == 0 {
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: Node %v has no CIDRs", node.Name)
 	}
-	_, cidr, err := net.ParseCIDR(cidrs[0])
-	if err != nil {
-		return fmt.Errorf("failed to parse string '%s' as a CIDR: %v", cidrs[0], err)
+	if _, err := netutils.ParseCIDRs(podCIDRs); err != nil {
+		return err
 	}
-	podCIDR := cidr.String()
 
-	if node.Spec.PodCIDR == podCIDR {
-		klog.V(4).Infof("Node %v already has allocated CIDR %v. It matches the proposed one.", node.Name, podCIDR)
+	skip := false
+
+	if len(podCIDRs) == 1 && node.Spec.PodCIDR == podCIDRs[0] {
+		skip = true
+		klog.V(4).Infof("Node %v already has allocated CIDR %v. It matches the proposed one.", node.Name, podCIDRs[0])
 		// We don't return here, in order to set the NetworkUnavailable condition later below.
-	} else {
+	} else if len(node.Spec.PodCIDRs) == len(podCIDRs) {
+		match := true
```

Comment:
* What happens if len(node.Spec.PodCIDR) > 1?
* Does the order have to match that we get from the alias IP metadata vs the IPs in node? Don't we have to put the IPv4 one first?
* Should we truncate the list of PodCIRDs from metadata server to 2, always?

Reply:
1. node.Spec.PodCIDR is a single string field and not an array. 
2. Yes. From a kubernetes point of view the order is important. node.Spec.PodCIDR == node.Spec.PodCIDRs[0]. This can be IPv4 or IPv6. For GKE dual stack clusters, this is going to be IPv4 followed by IPv6. In the future, if we support upgrading single stack IPv6 clusters to dual stack, then we will have to return an IPv6 address followed by IPv4 address.
3. Yes we should. It is an error to return more than 2 addresses.

Reply:
2. With the current design, we need to use node.Spec.PodCIDRs[0] for ipv4 and node.Spec.PodCIDRs[1] for ipv6 (if needed). We will make changes in AliasRangesByProviderID to match this behavior.
3. Yes, we should not get more than 2 ips. I have added a safety check to log and truncate to take the first 2 IPs. We can also add checks in AliasRangesByProviderID to avoid this.